src/components/cc-logs/cc-logs.js

```diff
@@ -0,0 +1,472 @@
+    };
+  };
+
+  constructor () {
+    super();
+
+    /** @type {boolean} Whether to force scroll to the bottom after adding logs */
+    this.follow = false;
+    /** @type {LogsState} The component state */
+    this.logs = { state: 'loading' };
+    /** @type {MessageRenderer} The custom message renderer */
+    this.messageRenderer = DEFAULT_MESSAGE_RENDERER;
+    /** @type {Object.<string, MetadataRenderer>} The custom renderers to use for displaying metadata */
+    this.metadataRenderers = {};
+    /** @type {TimestampDisplay} The timestamp display */
+    this.timestampDisplay = CcLogsComponent.DEFAULT_TIMESTAMP_DISPLAY;
+    /** @type {boolean} Whether to wrap long lines */
+    this.wrapLines = false;
+
+    /** @type {Ref<HTMLElement>} */
+    this._logsRef = createRef();
+    /** @type {boolean} */
+    this._pinned = this.follow;
+    /**
+     * @type {TimestampFormatter|null} instance of the formatter to use to format timestamp.
+     * It is maintained in sync with the `timestampDisplay` property.
+     * It is `null` when the `timestampDisplay` property is `hidden`.
+     */
+    this._timestampFormatter = this._resolveTimestampFormatter();
+  }
+
+  /* region private methods */
+
+  _resolveTimestampFormatter () {
+    return this.timestampDisplay === 'hidden' ? null : new TimestampFormatter(this.timestampDisplay);
+  }
+
+  /**
+   * When logs are appended very fast, the scroll event is fired very often.
+   * We sometimes need to inhibe the onScroll callback to achieve the user action.
+   */
+  _inhibiteScrollEventTemporarily () {
+    this._scrollEventInhibited = true;
+    setTimeout(() => {
+      this._scrollEventInhibited = false;
+    }, 50);
+  }
+
+  _refreshPinnedFromScrollPosition () {
+    const element = this._logsRef.value;
+    if (element != null) {
+      this._pinned = element.scrollTop + element.offsetHeight >= element.scrollHeight - 50;
+    }
+  }
+
+  _unpin () {
+    this._pinned = false;
+    this._inhibiteScrollEventTemporarily();
+  }
+
+  _pin () {
+    this._pinned = true;
+    this._inhibiteScrollEventTemporarily();
+  }
+
+  _onScroll () {
+    if (!this.follow || this._scrollEventInhibited) {
+      return;
+    }
+    this._refreshPinnedFromScrollPosition();
+  }
+
+  /**
+   * When logs are appended very fast, the scroll event is fired very often.
+   * We need to listen to the wheel event to force unpinning.
+   */
+  _onMouseWheel (e) {
+    if (!this.follow) {
+      return;
+    }
+    if (e.wheelDeltaY > 0) {
+      this._unpin();
+    }
+  }
+
+  /* endregion */
+
+  /* region public API */
+
+  scrollToBottom () {
+    if (this.follow) {
+      this._pin();
+    }
+    else {
+      const element = this._logsRef.value;
+      if (element != null) {
+        element.scrollTop = element.scrollHeight;
+      }
+    }
+  }
+
+  /* endregion */
+
+  firstUpdated (_changedProperties) {
+    if (_changedProperties.has('follow')) {
+      this._pinned = this.follow;
```

Comment:
Let say you have this:
```html
<cc-logs ?follow=${true}></cc-logs>
```

Here is how it goes:

1. constructor: `this.follow=false`, `this._pinned = false`
2. willUpdate: `this.follow=true`, `this._pinned = false`
3. firstUpdated: `this.follow=true`, `this._pinned = true`

If you remove the `firstUpdated`, the pinned will remain false.

Now you may wonder why in 2. the `willUpdate` doesn't move the `this._pinned` to `true`?

If you look closely at the `willUpdate` algorithm, it will move the `this._pinned` to `true` only if the scroll position is at the bottom (or near the bottom) which is not always the case at the time of the `firstUpdated`.

There is still a question pending: When the property `follow` changes from false to true, should we force pinning at the bottom even if the scroll position is not at the bottom.

Reply:
Thank you for this clarification, it makes sense! :pray: 

About the pending question, for the record, I think I like that it should force pinning at the bottom. But let's deal with it later, the current behavior is legit as well.